src/components/createOnyxContext.tsx

```diff
@@ -0,0 +1,85 @@
+import React, {ComponentType, ForwardRefExoticComponent, ForwardedRef, PropsWithoutRef, ReactNode, RefAttributes, createContext, forwardRef} from 'react';
+import {withOnyx} from 'react-native-onyx';
+import Str from 'expensify-common/lib/str';
+import getComponentDisplayName from '../libs/getComponentDisplayName';
+import {OnyxCollectionKey, OnyxKey, OnyxKeyValue, OnyxValues} from '../ONYXKEYS';
+
+type OnyxKeys = (OnyxKey | OnyxCollectionKey) & keyof OnyxValues;
+
+// Provider types
+type ProviderOnyxProps<TOnyxKey extends OnyxKeys> = Record<TOnyxKey, OnyxKeyValue<TOnyxKey>>;
+
+type ProviderProps = {
+    /** Rendered child component */
+    children: ReactNode;
+};
+
+type ProviderPropsWithOnyx<TOnyxKey extends OnyxKeys> = ProviderProps & ProviderOnyxProps<TOnyxKey>;
+
+// withOnyxKey types
+type WithOnyxKeyProps<TOnyxKey extends OnyxKeys, TNewOnyxKey extends string, TTransformedValue> = {
+    propName?: TOnyxKey | TNewOnyxKey;
+    // It's not possible to infer the type of props of the wrapped component, so we have to use `any` here
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    transformValue?: (value: OnyxKeyValue<TOnyxKey>, props: any) => TTransformedValue;
+};
+
+type WrapComponentWithConsumer<TNewOnyxKey extends string, TTransformedValue> = <TProps extends Record<TNewOnyxKey, TTransformedValue>, TRef>(
+    WrappedComponent: ComponentType<TProps & RefAttributes<TRef>>,
+) => ForwardRefExoticComponent<PropsWithoutRef<Omit<TProps, TNewOnyxKey>> & RefAttributes<TRef>>;
+
+type WithOnyxKey<TOnyxKey extends OnyxKeys> = <TNewOnyxKey extends string = TOnyxKey, TTransformedValue = OnyxKeyValue<TOnyxKey>>(
+    props?: WithOnyxKeyProps<TOnyxKey, TNewOnyxKey, TTransformedValue>,
+) => WrapComponentWithConsumer<TNewOnyxKey, TTransformedValue>;
+
+// createOnyxContext return type
+type CreateOnyxContext<TOnyxKey extends OnyxKeys> = [WithOnyxKey<TOnyxKey>, ComponentType<Omit<ProviderPropsWithOnyx<TOnyxKey>, TOnyxKey>>, React.Context<OnyxKeyValue<TOnyxKey>>];
+
+export default <TOnyxKey extends OnyxKeys>(onyxKeyName: TOnyxKey, defaultValue: OnyxKeyValue<TOnyxKey> = null): CreateOnyxContext<TOnyxKey> => {
```

Comment:
You've added `defaultValue = nul`l here, and I think it could mess up some uses. Could you please explain the necessity of having null here?

Reply:
After an internal discussion with @sebryu we found out that a default value is not needed at all as it was only used in this line:
```ts
const [withNetwork, NetworkProvider, NetworkContext] = createOnyxContext(ONYXKEYS.NETWORK, {});
```

I removed it completely and tested if it works well.